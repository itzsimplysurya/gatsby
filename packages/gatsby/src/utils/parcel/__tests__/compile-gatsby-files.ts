import { Parcel } from "@parcel/core"
import {
  constructBundler,
  compileGatsbyFiles,
  COMPILED_CACHE_DIR,
} from "../compile-gatsby-files"
import { siteMetadata } from "./fixtures/utils/site-metadata"
import { moreDataConfig } from "./fixtures/utils/more-data-config"
import { readFile, remove } from "fs-extra"

const dir = {
  js: `${__dirname}/fixtures/js`,
  ts: `${__dirname}/fixtures/ts`,
}

jest.mock(`@parcel/core`, () => {
  const parcelCore = jest.requireActual(`@parcel/core`)

  const { Parcel: OriginalParcel } = parcelCore

  class MockedParcel extends OriginalParcel {
    constructor(options) {
      super(options)
      this.options = options // Expose for assertions
    }
  }

  return {
    ...parcelCore,
    Parcel: MockedParcel,
  }
})

interface IMockedParcel extends Parcel {
  options: unknown
}

describe(`gatsby file compilation`, () => {
  describe(`constructBundler`, () => {
    it(`should construct Parcel relative to passed directory`, () => {
      const { options } = constructBundler(dir.js) as IMockedParcel

      expect(options).toMatchSnapshot({
        entries: `${dir.js}/gatsby-+(node|config).{ts,tsx,js}`,
        targets: {
          default: {
            distDir: `${dir.js}/${COMPILED_CACHE_DIR}`,
          },
        },
        cacheDir: `${dir.js}/.cache/.parcel-cache`,
      })
    })
  })

  describe(`compileGatsbyFiles`, () => {
    afterEach(async () => {
      for (const directory of [dir.js, dir.ts]) {
        await remove(`${directory}/.cache`)
      }
    })

    it(`should compile gatsby-config.js`, async () => {
      await compileGatsbyFiles(dir.js)

      const compiledGatsbyConfig = await readFile(
        `${dir.js}/.cache/compiled/gatsby-config.js`,
        `utf-8`
      )

      expect(compiledGatsbyConfig).toContain(siteMetadata.title)
      expect(compiledGatsbyConfig).toContain(siteMetadata.siteUrl)
      expect(compiledGatsbyConfig).toContain(moreDataConfig.options.name)
    })

    it(`should compile gatsby-config.ts`, async () => {
      await compileGatsbyFiles(dir.ts)

      const compiledGatsbyConfig = await readFile(
        `${dir.ts}/.cache/compiled/gatsby-config.js`,
        `utf-8`
      )

      expect(compiledGatsbyConfig).toContain(siteMetadata.title)
      expect(compiledGatsbyConfig).toContain(siteMetadata.siteUrl)
      expect(compiledGatsbyConfig).toContain(moreDataConfig.options.name)
    })

    it(`should compile gatsby-node.js`, async () => {
      await compileGatsbyFiles(dir.js)

      const compiledGatsbyNode = await readFile(
        `${dir.js}/.cache/compiled/gatsby-node.js`,
        `utf-8`
      )

      expect(compiledGatsbyNode).toContain(`I am working!`)
    })

    it(`should compile gatsby-node.ts`, async () => {
      await compileGatsbyFiles(dir.ts)

      const compiledGatsbyNode = await readFile(
        `${dir.ts}/.cache/compiled/gatsby-node.js`,
        `utf-8`
      )

      expect(compiledGatsbyNode).toContain(`I am working!`)
    })
  })
})
