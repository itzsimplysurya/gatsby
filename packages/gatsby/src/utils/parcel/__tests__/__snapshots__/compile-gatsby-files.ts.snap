// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gatsby file compilation constructBundler should construct Parcel relative to passed directory 1`] = `
Object {
  "cacheDir": "<PROJECT_ROOT>/packages/gatsby/src/utils/parcel/__tests__/fixtures/js/.cache/.parcel-cache",
  "defaultConfig": "@parcel/config-default",
  "entries": "<PROJECT_ROOT>/packages/gatsby/src/utils/parcel/__tests__/fixtures/js/gatsby-+(node|config).{ts,tsx,js}",
  "mode": "production",
  "targets": Object {
    "default": Object {
      "distDir": "<PROJECT_ROOT>/packages/gatsby/src/utils/parcel/__tests__/fixtures/js/.cache/compiled",
      "engines": Object {
        "node": ">= 14.15.0",
      },
      "includeNodeModules": false,
      "outputFormat": "commonjs",
      "sourceMap": false,
    },
  },
}
`;
